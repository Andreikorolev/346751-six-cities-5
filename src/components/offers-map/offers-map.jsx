import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import leaflet from "leaflet";

class OffersMap extends PureComponent {
  constructor(props) {
    super(props);

    this.mapContainer = React.createRef();
    this.map = null;
    this.markersGroup = null;
  }

  componentDidMount() {
    const {offers, activeCard, currentOffer} = this.props;
    const activeZoomControl = this.props.activeZoomControl;
    const activeScrollWheelZoom = this.props.activeScrollWheelZoom;

    const centerCity = [offers[0].city.location.latitude, offers[0].city.location.longitude];
    const zoom = offers[0].city.location.zoom;

    this.map = leaflet.map(this.mapContainer.current, {
      center: centerCity,
      zoom,
      zoomControl: activeZoomControl,
      scrollWheelZoom: activeScrollWheelZoom,
      marker: true
    });
    let icon;
    this.map.setView(centerCity, zoom);
    leaflet
    .tileLayer(`https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
      attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
    })
    .addTo(this.map);

    this.markersGroup = leaflet.layerGroup().addTo(this.map);

    if (currentOffer) {
      const offerCords = [currentOffer.location.latitude, currentOffer.location.longitude];

      icon = leaflet.icon({
        iconUrl: `img/pin-active.svg`,
        iconSize: [27, 39]
      });

      leaflet
        .marker(offerCords, {icon})
        .addTo(this.markersGroup);
    }

    offers.map((offer) => {
      const offerCords = [offer.location.latitude, offer.location.longitude];

      if (offer.id === activeCard) {
        icon = leaflet.icon({
          iconUrl: `img/pin-active.svg`,
          iconSize: [27, 39]
        });
      } else {
        icon = leaflet.icon({
          iconUrl: `img/pin.svg`,
          iconSize: [27, 39]
        });
      }

      leaflet
        .marker(offerCords, {icon})
        .addTo(this.markersGroup);
    });
  }

  render() {

    return (
      <section className="cities__map map">
        <div
          id="map"
          ref={this.mapContainer}
          style={{width: `100%`, height: `100%`}}
        ></div>
      </section>
    );
  }

  componentDidUpdate() {
    this.markersGroup.clearLayers();
    const {currentOffer, offers, activeCard} = this.props;
    const centerCity = [offers[0].city.location.latitude, offers[0].city.location.longitude];
    this.map.setView(centerCity);
    this.map.removeLayer(leaflet.marker);
    let icon;

    if (currentOffer) {
      const offerCords = [currentOffer.location.latitude, currentOffer.location.longitude];

      icon = leaflet.icon({
        iconUrl: `img/pin-active.svg`,
        iconSize: [27, 39]
      });

      leaflet
        .marker(offerCords, {icon})
        .addTo(this.markersGroup);
    }

    offers.map((offer) => {
      const offerCords = [offer.location.latitude, offer.location.longitude];

      if (offer.id === activeCard) {
        icon = leaflet.icon({
          iconUrl: `img/pin-active.svg`,
          iconSize: [27, 39]
        });
      } else {
        icon = leaflet.icon({
          iconUrl: `img/pin.svg`,
          iconSize: [27, 39]
        });
      }

      leaflet
        .marker(offerCords, {icon})
        .addTo(this.markersGroup);
    });
  }
}

OffersMap.propTypes = {
  currentOffer: PropTypes.oneOfType([
    PropTypes.object.isRequired,
    PropTypes.array.isRequired
  ]),
  offers: PropTypes.array.isRequired,
  activeCard: PropTypes.number.isRequired,
  activeZoomControl: PropTypes.bool.isRequired,
  activeScrollWheelZoom: PropTypes.bool.isRequired,
};

const mapStateToProps = ({PROCESS}) => ({
  activeCard: PROCESS.activeCard,
});

export {OffersMap};
export default connect(mapStateToProps)(OffersMap);
