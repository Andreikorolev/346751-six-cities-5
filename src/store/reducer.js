import {extend} from "../utils.js";
import {ActionType} from "./action";
import offers from "../mocs/offers";
import reviews from "../mocs/reviews";
import {Cities} from "../const";

const initialState = {
  city: Cities[0],
  offers: offers[Cities[0]],
  reviews,
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CITY_CHANGE:
      return extend(state, {
        city: action.payload,
      });

    case ActionType.GET_OFFERS_LIST:
      return extend(state, {
        offers: offers[action.payload],
      });
  }

  return state;
};

export {reducer};
