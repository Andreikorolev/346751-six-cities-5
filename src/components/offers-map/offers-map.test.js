import React from "react";
import renderer from "react-test-renderer";
import {OffersMap} from "./offers-map";
import {BrowserRouter} from "react-router-dom";
import {Provider} from "react-redux";
import {createStore} from "redux";
import reducer from "../../store/reducers/root-reducer";
import {offers} from "../../../mocks/mocks";

const store = createStore(reducer);

describe(`OffersMap render`, () => {
  it(`should render without current offer`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <BrowserRouter>
              <OffersMap
                offers={offers}
                currentOffer={null}
                activeZoomControl={true}
                activeScrollWheelZoom={false}
                activeCard={1}
              />
            </BrowserRouter>
          </Provider>,
          {
            createNodeMock: () => {
              return document.createElement(`div`);
            }
          }
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`should render with current offer`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <BrowserRouter>
              <OffersMap
                offers={offers}
                currentOffer={offers[0]}
                activeZoomControl={false}
                activeScrollWheelZoom={true}
                activeCard={1}
              />
            </BrowserRouter>
          </Provider>,
          {
            createNodeMock: () => {
              return document.createElement(`div`);
            }
          }
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
