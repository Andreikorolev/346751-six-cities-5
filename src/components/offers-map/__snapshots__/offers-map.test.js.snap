// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OffersMap render should render with current offer 1`] = `
<section
  className="cities__map map"
>
  <div
    id="map"
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  />
</section>
`;

exports[`OffersMap render should render without current offer 1`] = `
<section
  className="cities__map map"
>
  <div
    id="map"
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  />
</section>
`;
